crates/pallet-grandpa-finality-verifier/src/lib.rs

```diff
@@ -190,43 +188,41 @@ pub mod pallet {
             Error::<T>::InvalidBlock
         );
 
-        // if the block is the parent of the oldest known parent
-        // we update our state with its parent and import it
         let (oldest_known_parent_height, oldest_known_parent_hash) =
             OldestKnownParent::<T>::get(chain_id);
+
+        // if the block belongs to know parent, we import the block and progress backward
         if oldest_known_parent_height == number.into() {
             ensure!(
                 oldest_known_parent_hash == hash.into(),
                 Error::<T>::InvalidBlock
             );
 
-            // update our oldest known parent if we have not reached block 0 already
-            if oldest_known_parent_height > 0 {
-                OldestKnownParent::<T>::insert(
+            match number.checked_sub(&One::one()) {
+                // update state with new parent data
+                Some(parent_number) => OldestKnownParent::<T>::insert(
                     chain_id,
                     (
-                        oldest_known_parent_height - 1,
+                        parent_number.into(),
                         (*block.block.header.parent_hash()).into(),
                     ),
-                )
-            } else {
-                // chain has reached genesis.
-                // import this block and update the state so that further genesis block will fail when tried to import again
-                ensure!(
-                    !GenesisImported::<T>::get(chain_id),
-                    Error::<T>::InvalidBlock
-                );
-                GenesisImported::<T>::insert(chain_id, true);
+                ),
+                None => {
+                    // chain has reached genesis.
+                    // import genesis block once and update the state so that further genesis block imports will fail.
+                    ensure!(
+                        !GenesisImported::<T>::get(chain_id),
+                        Error::<T>::InvalidBlock
+                    );
+                    GenesisImported::<T>::insert(chain_id, true);
```

Comment:
I think `GenesisImported` is not necessary since we can figure it out from `oldest_known_parent_hash` (must have all zeroes) that we already read above.

Reply:
The whole `None` branch is in fact unnecessary as you'll not be able to submit block `0` again since `parent_hash` of the genesis block is all zeroes and you'll not find any other block that would hash to that, so validation will fail before getting here.

Reply:
fair!